refactor: split out `AutoFileReloader` from `QmlAutoReload`

Swiftb0y opened the pull request:
`AutoFileReload` is useful for almost all occasions where hot- reloading is needed. Thus it belongs into a separate class for easier reuse. Plus I added an extra RuntimeLoggingCategory so we don't have a category-less `qDebug` call.

Swiftb0y committed:
refactor: split out `AutoFileReloader` from `QmlAutoReload`

`AutoFileReload` is useful for almost all occasions where hot-
reloading is needed. Thus it belongs into a separate class for
easier reuse.

daschuer commented on src/util/autofilereloader.cpp:
```diff
@@ -0,0 +1,23 @@
+
+void AutoFileReloader::slotFileChanged(const QString& changedFile) {
+    qCDebug(m_loggingCategory) << "File" << changedFile << "has been changed.";
+    if (m_pathTriggeringReload.remove(changedFile)) {
```
Why is a changed file removed form the watcher?

Holzhaus replied:
See the comment in the header file. It is done to prevent double-reload caused by the different editor saving strategies (VSCode vs. vim/sed).

@Swiftb0y I think it should be pointed out in a comment that it's the callers responsibility to re-add the file once the reload was handled (in the QML case, this is done in the intercept call, after the changed file was parsed).

Swiftb0y replied:
Its in the header of `AutoFileReloader` as documentation for the `m_pathTriggeringReload` member.

daschuer commented on src/util/autofilereloader.cpp:
```diff
@@ -0,0 +1,23 @@
+void AutoFileReloader::slotFileChanged(const QString& changedFile) {
+    qCDebug(m_loggingCategory) << "File" << changedFile << "has been changed.";
+    if (m_pathTriggeringReload.remove(changedFile)) {
+        m_fileWatcher.removePath(changedFile);
```
removePath returns also true on success. What is the role for m_pathTriggeringReload?

Swiftb0y replied:
to ensure we don't emit a second signal in case of the append/truncate process (see the doc in the header).

acolombier replied:
@daschuer this is to specifically fix #13083.
When I implemented this auto reload, I wanted to make sure not to introduce the same limitation on QML.

acolombier replied:
iIRC, my initial implementation was exclusively relying on the file watcher to keep track of which files are being watched, but I can remember why I ended adding that extra array. I believe the initial PR/commits may have the explanation, I will try to dig it later, when I'm not AFK.

acolombier replied:
So the [original implementation](https://github.com/mixxxdj/mixxx/pull/12795/files#diff-f9cb0d49677f8a2d66f91e60c0d4c802bc51bdb25161a65118bbdf884a8d3d7c) only contained the QFileSystemWatcher, but we [had to add](https://github.com/mixxxdj/mixxx/pull/12844/files#diff-f9cb0d49677f8a2d66f91e60c0d4c802bc51bdb25161a65118bbdf884a8d3d7c) the array to desperately handle acknowledgement as [QFileSystemWatcher::removePath](https://doc.qt.io/qt-6/qfilesystemwatcher.html#removePath)'s returns appears to not be consistent against platforms.

acolombier reviewed (commented):
The code looks good, thanks for taking the time to do this well needed refactor.
I'll try to test that locally within the next few hours, days max. Please feel free to merge without my :+1:  if I'm taking too long to do so.

Swiftb0y commented:
Thanks for the testing. I'm not sure my tests tested both cases comprehensively.

acolombier reviewed (approved):
LGTM, thanks!

Swiftb0y committed:
docs: document callers responsibility to addPath after removal

This is needed because in some cases it may not be easy to
distinguish from a file changing and being removed.

daschuer reviewed (approved):
Ok, thank you. LGTM

Swiftb0y commented:
Thank you
